### ISSUE TITLE
Focus management between markers and popups

### ISSUE THREAD

**Issue by danallsop** [ScottLogic/jumpstart-Leaflet#347]
> **Migrated from:** https://github.com/Leaflet/Leaflet/issues/8115
> **Original author:** danallsop
> **Original created:** 2022-04-04T15:02:26Z
---

(Supersedes https://github.com/Leaflet/Leaflet/issues/2199.)

**Is your feature request related to a problem? Please describe.**

1. Popups are not reliably announced by screen readers after they've been opened

   (they _may_ be announced as a side effect of focus shift described in https://github.com/Leaflet/Leaflet/issues/8114, depending on the screen reader).

2. No logical focus order between elements (e.g. marker) and popups

   (this would be more reliably solved by https://github.com/Leaflet/Leaflet/issues/8113 for scenarios where the user is simply tabbing between elements and not necessarily opening/closing popups).

   [[#2199](https://github.com/Leaflet/Leaflet/issues/2199) (comment)](https://github.com/Leaflet/Leaflet/issues/2199#issuecomment-443924799):

   > ![leaflet-tab-issue 2018-12-03 at 16 35 17](https://user-images.githubusercontent.com/1072292/49410446-8573c500-f719-11e8-8a77-eeb080c1142a.gif)
   
   (http://bl.ocks.org/skorasaurus/9b2b96967335da91eac7676e815755d2)
  
3. Focus is not returned to the element that triggered it after a popup has been closed; the user cannot continue where they started before opening the popup.

**Describe the solution you'd like**

Screen readers announce content on focus delegation, and even though the modern best practice is to focus an element inside popups/dialogs Leaflet can't guarantee interactive content inside popups, so the suggestion is to focus the popup itself:

1. Set `tabindex="-1"` on `leaflet-popup-content-wrapper` ([to enable programmatic focus](https://webaim.org/techniques/keyboard/tabindex#negative-one))
2. Send focus to `leaflet-popup-content-wrapper` after the popup has been opened by the user
3. Send focus back to the source element (e.g. marker) after the popup has been closed (via <kbd>Esc</kbd> or the Close button)
4. https://github.com/Leaflet/Leaflet/issues/7968 (on `leaflet-popup`), can be decided later

**Additional context**

- Code demo: [Leaflet.js - Better Accessibility with Javascript: adjust the behaviour of popups so that focus is shifted to the popup content when it is opened](https://melmo.github.io/accessibility/code/dist/leaflet.html#:~:text=adjust%20the%20behaviour%20of%20popups%20so%20that%20focus%20is%20shifted%20to%20the%20popup%20content%20when%20it%20is%20opened)
- MapBox also does this: https://github.com/mapbox/mapbox-gl-js/issues/9662

**Comment by danallsop** [comment]
## Comments (11 total)

---
> **Comment by:** danallsop **on** 2022-04-19T18:38:46Z

Hey danallsop !
I would like to work upon this, if no one else is already on it.
Thanks!
---
> **Comment by:** danallsop **on** 2022-04-20T09:52:34Z

After closing a popup with the close button, the focus should go back to the marker of the popup
---
> **Comment by:** danallsop **on** 2022-04-20T16:31:15Z

danallsop 
Cool, I will get started on it.
---
> **Comment by:** danallsop **on** 2022-04-20T16:51:32Z

danallsop 
One doubt. I have installed the dependencies into my local machine. Can you tell me how can I run the app locally?
Thanks!
---
> **Comment by:** danallsop **on** 2022-04-20T16:56:47Z

You need to open a Terminal in the root of your project and run `npm run watch` which generates the `leaflet-src.js` file in the `dist` folder. For testing you can open `/debug/map/map-mobile.html`
---
> **Comment by:** danallsop **on** 2022-04-21T14:51:36Z

Thanks for creating a new issue; I updated the test case that y'all can use; uses leaflet 1.8.0; it's now at http://skorasaurus.github.io/maptemplates/leaflet-test-8115.html to reflect the new issue number. 
---
> **Comment by:** danallsop **on** 2022-06-15T15:07:39Z

Hi, I have a question, How did you add a listener on the button in the popup?
---
> **Comment by:** danallsop **on** 2022-06-15T15:10:46Z

danallsop please look into: https://stackoverflow.com/a/71176793/8283938
---
> **Comment by:** danallsop **on** 2022-08-08T15:47:50Z

This might be slightly off-topic here, feel free to move to a new issue...

Does the Open UI Pop Up API proposal meet the requirements for web maps / Leaflet?
https://open-ui.org/components/popup.research.explainer

(File new issues for open-ui here https://github.com/openui/open-ui/issues )
---
> **Comment by:** danallsop **on** 2023-09-11T15:16:17Z

Two points:

1. If a natural tab-order is implemented (#345), then probably it would be best to follow the simple [disclosure-widget pattern](https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/), which uses `aria-expanded` but does _not_ involve automatically moving focus to the newly opened region (also means no `tabindex="-1"`).
  
    It might be nice in this case to allow the user to _choose_ what `role` to use for the Popup. Often there's no need for a `role` at all with the disclosure-pattern, but since the Popup's got a close-button I think it should have one. In some circumstances `role="dialog"` might be a good choice (with `aria-modal="false"`), but a simple `role="group"` would often suffice, and may be better sometimes. For example, if the map-Popup's got a button that opens a modal dialog, you'd probably want to avoid giving the map-Popup `role="dialog"`.

2. At the risk of stating the obvious: it's important that focus be returned to the Marker on Popup-close _only when the Popup was closed from within_. We don't want focus going back to the Marker if the Popup was closed for some other reason (e.g., user clicked a different Marker).
---
> **Comment by:** danallsop **on** 2025-04-25T14:16:01Z

Could this be solved by the [popover api](https://developer.mozilla.org/en-US/docs/Web/API/Popover_API/Using)?

**Comment by snakefood3232** [comment]
The focus management issue where popups aren't announced and focus isn't returned to the marker on close is critical for accessibility. I'd handle this by adding the `tabindex="-1"` to the content wrapper and managing focus events on open/close as you described. Can have a PR up within a day. I've implemented similar accessible focus trapping in React map components.

**Comment by snakefood3232** [comment]
Focus management for screen readers is clearly broken when popups open without announcing content. I'd handle this by adding tabindex="-1" to leaflet-popup-content-wrapper and managing focus programmatically on open/close events. Can have a PR up in a day to implement the described solution. I've worked on similar accessible map components in React/TypeScript before.
